fastestimator/op/tensorop/model/update.py

```diff
@@ -74,22 +111,132 @@ def fe_retain_graph(self, retain: Optional[bool] = None) -> Optional[bool]:
+        """
+        if self.framework == "tf":
+            gradients = get_gradient(loss, self.model.trainable_variables, tape=tape)
+
+        else:  # self.framework == "torch"
+            trainable_params = [p for p in self.model.parameters() if p.requires_grad]
+            try:
+                gradients = get_gradient(loss, trainable_params, retain_graph=self.retain_graph)
+            except RuntimeError as err:
+                if err.args and isinstance(err.args[0], str) and err.args[0].startswith(
+                        'one of the variables needed for gradient computation has been modified by an inplace operation'
+                ):
+                    raise RuntimeError(
+                        "When computing gradients for '{}', some variables it relied on during the forward pass had"
+                        " been updated. Consider setting defer=True in earlier UpdateOps related to models which "
+                        "interact with this one.".format(self.model.model_name))
+                raise err
+
+        return gradients
+
+    def _gradient_postprocess(self, gradients: Union[Tensor, List[Tensor]]) -> Union[Tensor, List[Tensor]]:
+        """Gradient postprocess for multi-GPU and mixed-precision training.
+
+        Args:
+            gradients: Unprocessed gradients.
+
+        Returns:
+            Processed gradients.
+        """
+        if self.framework == "tf":
+            if self.gradients is not None:  # when user provide gradients
+                strategy = tf.distribute.get_strategy()
+                # for multi-gpu training, the gradient will be combined by sum, normalize the gradient
+                if isinstance(strategy, tf.distribute.MirroredStrategy):
+                    gradients = gradients / strategy.num_replicas_in_sync
+
+            if self.model.mixed_precision:
+                # scale down gradient to balance scale-up loss
+                gradients = self.model.current_optimizer.get_unscaled_gradients(gradients)
+
+        return gradients
+
+    def _merge_grad_update(self,
+                           gradients: Union[Tensor, List[Tensor]],
+                           deferred: Optional[Dict[str, List[Callable[[], None]]]] = None) -> None:
+        """Accumulate gradients and update the model at certain frequency of invocation.
+
+        Args:
+            gradients: Input gradients.
+            deferred: A dictionary in which model update functions are stored.
+        """
+
+        # add current gradient to the cumulative gradient
+        for gs, g in zip(self.grad_sum, gradients):
+            self._assign_add(gs, g)
+
+        self._assign_add(self.step, 1)
+
+        if self.step % self.merge_grad == 0:
+            update_model(model=self.model, gradients=self.grad_sum, defer=self.defer, deferred=deferred)
+            for gs in self.grad_sum:
+                self._assign_add(gs, -gs)  # zero the gradient in place
+
+    def _assign_add(self, a, b):
```

Comment:
typing. but also, maybe we should make an assign backend function since both this and super loss are essentially implementing the same thing.

Reply:
Ya we can do that

Reply:
I just checked and found there are some differences. The superloss uses "assign" but here it uses "assign_add".

Reply:
I didn't make it backend function but I added the typing.